### ISSUE TITLE
Field 'xxx': type mismatch. Was INTEGER, expected STRING

### ISSUE THREAD

**Issue by diegomontoya** [realm/realm-java#1610]
Related to https://github.com/realm/realm-java/issues/1319

We are still seeing this issue with 10-12 snapshot. 

```
java.lang.IllegalArgumentException: Field 'status': type mismatch. Was INTEGER, expected STRING.
       at io.realm.RealmQuery.getColumnIndices(RealmQuery.java:167)
       at io.realm.RealmQuery.equalTo(RealmQuery.java:305)
```

Crashed point with query 

```
realm.where(MyRealm.class).equalTo("status", new Integer(1)).findAll()
```

The column in question, 'status' is an "int". 

```
    @PrimaryKey
    private long a;

    @Index
    private int b;

    @Index
    private int c;

    @Required
    private String d;

    @Index
    private int status;
```

Can a db corruption cause column def corruption? If so, it may be time to move db def to a separate file to ensure such things cannot happen.

**Comment by cmelchior** [comment]
It is very unlikely to be DB corruption, it is more likely our translation of your fields to our database types . We cache a mapping between your field names and database types when opening a Realm, this cache is used when creating the queries, so if something changed after we created the cache and we didn't clear it properly you would get an error like the one you are seeing. You can see the field just before `status` is a String.

Do you run any migrations on your database? Also are you able to reproduce this?

**Comment by diegomontoya** [comment]
There is no migration defined for this realm instance/db. We are using separate realm db for each unique table so this db only has 1 table with no migration. We cannot reproduce this bug on our devices but the occurrence of the bug is relatively high at around 0.3% of users.

**Comment by diegomontoya** [comment]
Sidebar

RealmQuery.java Line 161

columns.get(fieldName) called potentially 3 times. 

```java
            if (columns.get(fieldName) == null) {
                throw new IllegalArgumentException(String.format("Field '%s' does not exist.", fieldName));
            }

            ColumnType tableColumnType = table.getColumnType(columns.get(fieldName));
            if (fieldType != null && fieldType != tableColumnType) {
                throw new IllegalArgumentException(String.format("Field '%s': type mismatch. Was %s, expected %s.",
                        fieldName, fieldType, tableColumnType
                ));
            }
            return new long[] {columns.get(fieldName)};
```

Also, why does getColumnType need to call native jni to get column type every single time and not cached? Strange this has to be done every time since column index is cached but column data type is not. 

```java
    @Override
    public ColumnType getColumnType(long columnIndex) {
        return ColumnType.fromNativeValue(nativeGetColumnType(nativePtr, columnIndex));
    }
```

**Comment by cmelchior** [comment]
Just to be sure. You have multiple Realm files each with their own configuration? Do you perform a migration on any of those?  Also can you post the configuration for each of your Realm instances?

**Comment by diegomontoya** [comment]
Correct. Each realm file contains only 1 table and each with its own configuration. Of the dozen or so realm configs, we have 2 configs with migration defined but the chance of those migration is actually 0% since they can't be triggered with our current code. 

I think you are asking if any other realm config/instance might have performed migration that may have affected another realm config/instance. My answer is I do not believe any migration was ever performed.

**Comment by cmelchior** [comment]
@diegomontoya You can send them to help@realm.io. Thanks a lot.

**Comment by diegomontoya** [comment]
New crash variation at exact same location. 

```java
java.lang.IllegalArgumentException: Field 'status': type mismatch. Was INTEGER, expected BOOLEAN.
```

@cmelchior Will send you the config/modules/models.

**Comment by bmunkholm** [comment]
@diegomontoya Are you btw using a SNAPSHOT release in your release to users? Although we actually highly appreciate developers using our SNAPSHOT releases, we highly recommend to only use official releases for your app releases.

**Comment by cmelchior** [comment]
Hi @diegomontoya. I have discovered a problem in our Query API. If there is hash collisions between your model class names, we can reproduce the behaviour you are seeing. So if you can send a list of all your model classes/modules/configurations that would be really helpfull.

**Comment by diegomontoya** [comment]
@cmelchior Just sent you the files.

**Comment by cmelchior** [comment]
it is most likely only Java HashMap/HashSet usage. I'm still investigating, but will update this issue once I know more.

**Comment by cmelchior** [comment]
@diegomontoya Are you manually calling `Realm.migrateRealm()` anywhere?

**Comment by diegomontoya** [comment]
@cmelchior  No. Not calling migration manually for any of the realms.

**Comment by cmelchior** [comment]
Hi @diegomontoya 
After some investigation it turned out my first idea was wrong, it was due to a mistake in my unit test. So I  am still searching for a reason that his could happen.

**Comment by cmelchior** [comment]
Hi @diegomontoya 
We added a fix to another caching issue we fond that might effect this. Can you please try to see if you can still reproduce this error with our `0.84.1-SNAPSHOT`?

**Comment by cmelchior** [comment]
@diegomontoya We just released the official `0.84.1` as well, so you can also test on that.

**Comment by diegomontoya** [comment]
@cmelchior Thanks. Should know in a few days if 0.84.1 has resolved this issue on my end.

**Comment by diegomontoya** [comment]
@cmelchior I believe this bug is fixed with the recent commits as of 0.84.1

**Comment by cmelchior** [comment]
Good to hear. I'll close this issue then. Feel free to reopen if you discover otherwise.
